api/http/handlerfuncs_export.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package http
+
+import (
+	"net/http"
+
+	"github.com/fxamacker/cbor/v2"
+
+	"github.com/sourcenetwork/defradb/client"
+)
+
+func exportHandler(rw http.ResponseWriter, req *http.Request) {
+	db, err := dbFromContext(req.Context())
+	if err != nil {
+		handleErr(req.Context(), rw, err, http.StatusInternalServerError)
+		return
+	}
+
+	showDeletedParam := req.URL.Query().Get("showDeleted")
+	showDeleted := showDeletedParam == "true"
+	colsList := req.URL.Query()["collections"]
+
+	// map of collection name pointing to slices of documents.
+	docs := make(map[string][]map[string]any)
+	getDocs := func(rw http.ResponseWriter, req *http.Request, col client.Collection) (docs []map[string]any, ok bool) {
+		keysCh, err := col.GetAllDocKeys(req.Context())
+		if err != nil {
+			handleErr(req.Context(), rw, err, http.StatusInternalServerError)
+			return nil, false
+		}
+
+		for key := range keysCh {
+			doc, err := col.Get(req.Context(), key.Key, showDeleted)
+			if err != nil {
```

Comment:
todo: This looks like it might error should the document have been deleted (with `showDeleted == false`) between here and the retrieval of the key.  It also looks as if it will error out on any transaction conflict.

I don't think either of those things should result in the entire export being cancelled. Such behaviour might make it highly impractical to ever export some production databases.

Reply:
I can change it to use a single transaction for the whole export which would be a better approach anyways.

Reply:
I think a single transaction would be even worse, you'd be as good as locking the whole database for the duration of the export, and any transaction already in progress would result in a conflict (also cancelling the whole export)

Reply:
Why would the export be canceled? We can't get a transaction conflict on read. Thinking about it more, why would a deleted document cause any problem if the getting of the documents is done on a single transaction?

Reply:
> We can't get a transaction conflict on read.

We never get them on read? Only update?  Is not a problem them, from that perspective.  The very large transaction might cause some memory problems of its own though - worth testing that.

Delete would never be an issue with a single txn.

Would also likely run into the problem that a single txn would complicate the solving of the memory issue however, as we currently have no means to preserve a txn between remove calls (i.e. the whole server side op would be forced to copy everything into memory until/unless we implement that too).